statesync/snapshots.go

```diff
@@ -0,0 +1,260 @@
+}
+
+// newSnapshotPool creates a new snapshot pool.
+func newSnapshotPool(stateSource StateSource) *snapshotPool {
+	return &snapshotPool{
+		stateSource:       stateSource,
+		snapshots:         make(map[snapshotHash]*snapshot),
+		snapshotPeers:     make(map[snapshotHash]map[p2p.ID]p2p.Peer),
+		formatIndex:       make(map[uint32]map[snapshotHash]bool),
+		heightIndex:       make(map[uint64]map[snapshotHash]bool),
+		peerIndex:         make(map[p2p.ID]map[snapshotHash]bool),
+		formatBlacklist:   make(map[uint32]bool),
+		heightBlacklist:   make(map[uint64]bool),
+		snapshotBlacklist: make(map[snapshotHash]bool),
+	}
+}
+
+// Add adds a snapshot to the pool, unless the peer has already sent recentSnapshots snapshots. It
+// returns true if this was a new, non-blacklisted snapshot. The snapshot height is verified using
+// the light client, and the expected app hash is set for the snapshot.
+func (p *snapshotPool) Add(peer p2p.Peer, snapshot *snapshot) (bool, error) {
+	appHash, err := p.stateSource.AppHash(snapshot.Height)
+	if err != nil {
+		return false, err
+	}
+	snapshot.trustedAppHash = appHash
+	hash := snapshot.Hash()
+
+	p.Lock()
+	defer p.Unlock()
+
+	switch {
+	case p.formatBlacklist[snapshot.Format]:
+		return false, nil
+	case p.heightBlacklist[snapshot.Height]:
+		return false, nil
+	case p.snapshotBlacklist[hash]:
+		return false, nil
+	case p.snapshots[hash] != nil:
+		p.snapshotPeers[hash][peer.ID()] = peer
+		return false, nil
+	case len(p.peerIndex[peer.ID()]) >= recentSnapshots:
+		return false, nil
+	}
+
+	p.snapshots[hash] = snapshot
+
+	if p.snapshotPeers[hash] == nil {
+		p.snapshotPeers[hash] = make(map[p2p.ID]p2p.Peer)
+	}
+	p.snapshotPeers[hash][peer.ID()] = peer
+
+	if p.peerIndex[peer.ID()] == nil {
+		p.peerIndex[peer.ID()] = make(map[snapshotHash]bool)
+	}
+	p.peerIndex[peer.ID()][hash] = true
+
+	if p.formatIndex[snapshot.Format] == nil {
+		p.formatIndex[snapshot.Format] = make(map[snapshotHash]bool)
+	}
+	p.formatIndex[snapshot.Format][hash] = true
+
+	if p.heightIndex[snapshot.Height] == nil {
+		p.heightIndex[snapshot.Height] = make(map[snapshotHash]bool)
+	}
+	p.heightIndex[snapshot.Height][hash] = true
+
+	return true, nil
+}
+
+// Best returns the "best" currently known snapshot, if any.
+func (p *snapshotPool) Best() *snapshot {
+	ranked := p.Ranked()
+	if len(ranked) == 0 {
+		return nil
+	}
+	return ranked[0]
+}
+
+// GetPeer returns a random peer for a snapshot, if any.
+func (p *snapshotPool) GetPeer(snapshot *snapshot) p2p.Peer {
+	peers := p.GetPeers(snapshot)
+	if len(peers) == 0 {
+		return nil
+	}
+	return peers[rand.Intn(len(peers))]
+}
+
+// GetPeers returns the peers for a snapshot.
+func (p *snapshotPool) GetPeers(snapshot *snapshot) []p2p.Peer {
+	hash := snapshot.Hash()
+	p.Lock()
+	defer p.Unlock()
+
+	peers := make([]p2p.Peer, 0, len(p.snapshotPeers[hash]))
+	for _, peer := range p.snapshotPeers[hash] {
+		peers = append(peers, peer)
+	}
+	sort.Slice(peers, func(a int, b int) bool {
```

Comment:
why sort them?

Reply:
For testability. Otherwise their order will be random, and tests will break randomly.